src/runtime/metadata_module.cc

```diff
@@ -0,0 +1,220 @@
+        init(md);
+        break;
+      }
+    }
+  }
+
+  void SaveToBinary(dmlc::Stream* stream) final {
+    std::vector<std::string> variables;
+    std::vector<NDArray> metadata;
+    for (const auto& it : metadata_) {
+      String var_name = it.first;
+      variables.push_back(var_name);
+      metadata.push_back(it.second);
+    }
+
+    // Save all variables in the function.
+    stream->Write(variables);
+    // Save all constant data.
+    uint64_t sz = static_cast<uint64_t>(metadata.size());
+    stream->Write(sz);
+    for (uint64_t i = 0; i < sz; i++) {
+      metadata[i].Save(stream);
+    }
+
+    // Save the symbol to list of required constant variables mapping
+    std::vector<std::string> symbols;
+    std::vector<std::vector<std::string>> const_vars;
+    for (const auto& it : sym_vars_) {
+      symbols.push_back(it.first);
+      const_vars.push_back(it.second);
+    }
+
+    stream->Write(symbols);
+    sz = static_cast<uint64_t>(sym_vars_.size());
+    stream->Write(sz);
+    for (uint64_t i = 0; i < sz; i++) {
+      stream->Write(const_vars[i]);
+    }
+  }
+
+  static Module LoadFromBinary(void* strm) {
+    dmlc::Stream* stream = static_cast<dmlc::Stream*>(strm);
+
+    // Load the variables.
+    std::vector<std::string> variables;
+    CHECK(stream->Read(&variables)) << "Loading variables failed";
+    uint64_t sz;
+    CHECK(stream->Read(&sz, sizeof(sz))) << "Loading medata size failed";
+    CHECK_EQ(static_cast<size_t>(sz), variables.size())
+        << "The number of variables and ndarray counts must match";
+    // Load the list of ndarray.
+    std::vector<NDArray> arrays;
+    for (uint64_t i = 0; i < sz; i++) {
+      NDArray temp;
+      temp.Load(stream);
+      arrays.push_back(temp);
+    }
+
+    std::unordered_map<std::string, NDArray> metadata;
+    for (size_t i = 0; i < variables.size(); i++) {
+      CHECK_EQ(metadata.count(variables[i]), 0U);
+      metadata[variables[i]] = arrays[i];
+    }
+
+    // Load the symbol to list of required constant variables mapping
+    std::vector<std::string> symbols;
+    CHECK(stream->Read(&symbols)) << "Loading symbols failed";
+    CHECK(stream->Read(&sz, sizeof(sz))) << "Loading number of symbols failed";
+    CHECK_EQ(static_cast<size_t>(sz), symbols.size());
+    std::vector<std::vector<std::string>> const_vars;
+    for (uint64_t i = 0; i < sz; i++) {
+      std::vector<std::string> vars;
+      CHECK(stream->Read(&vars)) << "Loading const variables failed";
+      const_vars.push_back(vars);
+    }
+
+    std::unordered_map<std::string, std::vector<std::string>> sym_vars;
+    for (uint64_t i = 0; i < sz; i++) {
+      sym_vars[symbols[i]] = const_vars[i];
+    }
+
+    auto n = make_object<MetadataModuleNode>(metadata, sym_vars);
+    return Module(n);
+  }
+
+ private:
+  /*!
+   * \brief Record if a module is initialized. It is needed by imported
+   * modules using execution engine.
+   */
+  std::unordered_map<std::string, bool> initialized_;
```

Comment:
`unordered_set` fits better for this case, but I suppose `unordered_map` will be replaced with `runtime::Map` in the future so it should be fine to keep it.